src/main/java/org/ibp/api/java/impl/middleware/dataset/DatasetServiceImpl.java

```diff
@@ -844,4 +863,90 @@ private void addLocationIdVariable(final List<MeasurementVariable> environmentDe
 		locationIdVariable.setName(LOCATION_ID_VARIABLE_NAME);
 		environmentDetailAndConditionVariables.add(0, locationIdVariable);
 	}
+
+	@Override
+	public void replaceObservationUnitsEntry(final int studyId, final int datasetId, final ObservationUnitEntryReplaceRequest request) {
+		this.studyValidator.validate(studyId, true);
+		this.datasetValidator.validateDataset(studyId, datasetId);
+		this.datasetValidator.validatePlotDatasetType(datasetId);
+		final BindingResult errors = new MapBindingResult(new HashMap<String, String>(), Integer.class.getName());
+		if (this.studyService.studyHasGivenDatasetType(studyId, DatasetTypeEnum.MEANS_DATA.getId())) {
+			errors.reject("study.has.means.dataset");
+			throw new ApiRequestValidationException(errors.getAllErrors());
+		}
+
+		BaseValidator.checkNotNull(request, "param.null", new String[] {"request"});
+		BaseValidator.checkNotNull(request.getSearchRequest(), "param.null", new String[] {"searchRequest"});
+		BaseValidator.checkNotNull(request.getEntryId(), "param.null", new String[] {"entryId"});
+
+		this.inventoryCommonValidator.validateSearchCompositeDto(request.getSearchRequest(), errors);
+
+		studyValidator.validateStudyContainsEntry(studyId, request.getEntryId());
+
+		if (this.studyService.hasCrossesOrSelections(studyId)) {
+			errors.reject("study.has.crosses.or.selections");
+			throw new ApiRequestValidationException(errors.getAllErrors());
+		}
+
+		this.processSearchComposite(request.getSearchRequest());
+
+		// observation units
```

Comment:
Extract to 2 methods: `validateObservationUnitsHasNoSamples` and `validateObservationUnitsHasNoTransactions` (or something similarly named) to make the main method easier to understand/read